src/Components/Components/src/PersistentComponentState.cs

```diff
@@ -105,27 +94,30 @@ public bool TryTakePersistedState(string key, [MaybeNullWhen(false)] out byte[]?
         }
 
         /// <summary>
-        /// Persists the serialized state <paramref name="value"/> for the given <paramref name="key"/>.
+        /// Persists the serialized state <paramref name="valueWriter"/> for the given <paramref name="key"/>.
         /// </summary>
         /// <param name="key">The key to use to persist the state.</param>
-        /// <param name="value">The state to persist.</param>
-        public void PersistState(string key, byte[] value)
+        /// <param name="valueWriter">The state to persist.</param>
+        public void Persist(string key, Action<IBufferWriter<byte>> valueWriter)
         {
             if (key is null)
             {
                 throw new ArgumentNullException(nameof(key));
             }
 
-            if (value is null)
+            if (valueWriter is null)
             {
-                throw new ArgumentNullException(nameof(value));
+                throw new ArgumentNullException(nameof(valueWriter));
             }
 
             if (_currentState.ContainsKey(key))
             {
                 throw new ArgumentException($"There is already a persisted object under the same key '{key}'");
             }
-            _currentState.Add(key, value);
+
+            var writer = new PooledByteBufferWriter();
```

Comment:
This gives me pause. So for every separate item being persisted (of which there could be thousands), we rent a new buffer from a pool? It seems odd, but maybe I'm misunderstanding. I would have expected a single writer to be adequate for the whole store.

Reply:
See the perf numbers for reference. I'm not optimizing for thousand separate items being persisted. That will be part of the docs guidance. Iex, don't save each individual row, but save the table data instead.

A single blob works well for prerendering, but might not work well for other cases like redis, sql, etc. Stores have the ability to define how they serialize the store data.

Reply:
I've also added code to enforce now that you can only call PersistState within the context of an OnPersisting callback, which means that this buffer is rented and released in the shortest amount of time possible.

For scenarios I would refer to the perf numbers, however as with anything, every system is going to have a limit and I don't expect us to efficiently support 10.000 individual items. Around 100 is already quite a high number